zkevm-circuits/src/evm_circuit/util/common_gadget/tx_eip2930.rs

```diff
@@ -0,0 +1,134 @@
+use super::{CachedRegion, Cell};
+use crate::{
+    evm_circuit::util::{
+        constraint_builder::EVMConstraintBuilder, math_gadget::IsEqualGadget, select,
+    },
+    table::TxFieldTag,
+    util::Expr,
+    witness::Transaction,
+};
+use bus_mapping::circuit_input_builder::CopyDataType;
+use eth_types::{
+    geth_types::{access_list_address_and_storage_key_sizes, TxType},
+    Field,
+};
+use halo2_proofs::{
+    circuit::Value,
+    plonk::{Error, Expression},
+};
+
+/// Transaction EIP-2930 gadget to handle optional access-list
+#[derive(Clone, Debug)]
+pub(crate) struct TxEip2930Gadget<F> {
+    is_eip2930_tx: IsEqualGadget<F>,
+    access_list_address_len: Cell<F>,
+    access_list_storage_key_len: Cell<F>,
+    access_list_gas_cost: Cell<F>,
+}
+
+impl<F: Field> TxEip2930Gadget<F> {
+    pub(crate) fn construct(
+        cb: &mut EVMConstraintBuilder<F>,
+        tx_id: Expression<F>,
+        tx_type: Expression<F>,
+    ) -> Self {
+        let is_eip2930_tx = IsEqualGadget::construct(cb, tx_type, (TxType::Eip2930 as u64).expr());
+
+        let [access_list_address_len, access_list_storage_key_len, access_list_gas_cost] = cb
+            .condition(is_eip2930_tx.expr(), |cb| {
+                let [address_len, storage_key_len, gas_cost] = [
+                    TxFieldTag::AccessListAddressesLen,
+                    TxFieldTag::AccessListStorageKeysLen,
+                    TxFieldTag::AccessListGasCost,
+                ]
+                .map(|field_tag| cb.tx_context(tx_id.expr(), field_tag, None));
+
+                // Let copy-circuit to write the tx-table's access list addresses into rw-table.
+                cb.copy_table_lookup(
```

Comment:
Comment related to the bus-mapping `state.push_copy`. If you push the event conditionally (whether the access list is non-empty), then you will need an additional condition here so that the lookup does not fail.